torch/csrc/api/src/nn/modules/conv.cpp

```diff
@@ -20,119 +25,163 @@ ConvImpl<D, Derived>::ConvImpl(const ConvOptions<D>& options_)
 
 template <size_t D, typename Derived>
 void ConvImpl<D, Derived>::reset() {
-  if (!options.transposed()) {
-    for (auto pad : *options.output_padding()) {
-      TORCH_CHECK(
-          pad == 0, "Only transposed convolutions support output padding!");
-    }
-  }
+  TORCH_CHECK(options.in_channels().has_value(), "in_channels needs to be specified in options");
+  TORCH_CHECK(options.out_channels().has_value(), "out_channels needs to be specified in options");
+  TORCH_CHECK(options.kernel_size().has_value(), "kernel_size needs to be specified in options");
+
+  TORCH_CHECK(
+    *options.in_channels() % options.groups() == 0,
+    "in_channels must be divisible by groups");
+  TORCH_CHECK(
+    *options.out_channels() % options.groups() == 0,
+    "out_channels must be divisible by groups");
 
-  std::vector<int64_t> weights_size;
   if (options.transposed()) {
-    weights_size.push_back(options.input_channels());
-    weights_size.push_back(options.output_channels() / options.groups());
+    std::vector<int64_t> weight_sizes = {
+      *options.in_channels(),
+      *options.out_channels() / options.groups()};
+    weight_sizes.insert(weight_sizes.end(), (*options.kernel_size().value()).begin(), (*options.kernel_size().value()).end());
+    weight = this->register_parameter(
+      "weight",
+      torch::empty(weight_sizes));
   } else {
-    weights_size.push_back(options.output_channels());
-    weights_size.push_back(options.input_channels() / options.groups());
+    std::vector<int64_t> weight_sizes = {
+      *options.out_channels(),
+      *options.in_channels() / options.groups()};
+    weight_sizes.insert(weight_sizes.end(), (*options.kernel_size().value()).begin(), (*options.kernel_size().value()).end());
+    weight = this->register_parameter(
+      "weight",
+      torch::empty(weight_sizes));
   }
-  weights_size.insert(
-      weights_size.end(),
-      options.kernel_size()->begin(),
-      options.kernel_size()->end());
-  AT_ASSERT(weights_size.size() == 2 + options.kernel_size()->size());
-
-  weight = this->register_parameter("weight", torch::empty(weights_size));
-  if (options.with_bias()) {
-    bias = this->register_parameter(
-        "bias", torch::empty(options.output_channels()));
+
+  if (options.bias()) {
+    bias = this->register_parameter("bias", torch::empty({*options.out_channels()}));
+  } else {
+    this->register_parameter("bias", Tensor(), /*requires_grad=*/false);
   }
 
-  const auto number_of_features = std::accumulate(
-      options.kernel_size()->begin(),
-      options.kernel_size()->end(),
-      options.input_channels(),
-      std::multiplies<int64_t>{});
-  const auto stdv = 1.0 / std::sqrt(number_of_features);
-  NoGradGuard no_grad;
-  for (auto& p : this->parameters()) {
-    p.uniform_(-stdv, stdv);
+  init::kaiming_uniform_(weight, /*a=*/std::sqrt(5));
```

Comment:
```suggestion
  init::kaiming_uniform_(weight, /*a=*/std::sqrt(5));  // NOLINT(cppcoreguidelines-avoid-magic-numbers)
```
`/*a=*/` does not give any useful information

Reply:
Thanks for the catch! I added `// NOLINT(cppcoreguidelines-avoid-magic-numbers)`. For `/*a=*/`, I was trying to mimic the Python version as much as possible, and the Python version does this:
```python
init.kaiming_uniform_(self.weight, a=math.sqrt(5))
```
I think it would be good to highlight that the second argument is for `a` to more closely match the Python version, which makes reviewing for parity easier.